backends/cdc-postgres/read.go

```diff
@@ -0,0 +1,150 @@
+
+	handler := func(m pgoutput.Message, _ uint64) error {
+		switch v := m.(type) {
+		case pgoutput.Begin:
+			changeRecord.Timestamp = v.Timestamp.UTC().UnixNano()
+			changeRecord.XID = v.XID
+		case pgoutput.Commit:
+			changeRecord.LSN = pgx.FormatLSN(v.LSN)
+
+			// Advance LSN so we do not read the same messages on re-connect
+			sub.AdvanceLSN(v.LSN + 1)
+
+			output, _ := json.MarshalIndent(changeRecord, "", "  ")
+			p.Printer.Print(string(output))
+
+			changeRecord = &types.ChangeRecord{}
+			changeRecord.Changes = make([]*types.Change, 0)
+		case pgoutput.Relation:
+			set.Add(v)
+		case pgoutput.Insert:
+			changeSet, ok := set.Get(v.RelationID)
+			if !ok {
+				return errors.New("relation not found")
+			}
+			values, err := set.Values(v.RelationID, v.Row)
+			if err != nil {
+				return fmt.Errorf("error parsing values: %s", err)
+			}
+
+			change := &types.Change{
+				Operation: "insert",
+				Table:     changeSet.Name,
+				Fields:    make(map[string]interface{}, 0),
+			}
+			for name, value := range values {
+				change.Fields[name] = value.Get()
+			}
+			changeRecord.Changes = append(changeRecord.Changes, change)
+		case pgoutput.Update:
+			changeSet, ok := set.Get(v.RelationID)
+			if !ok {
+				return errors.New("relation not found")
+			}
+			values, err := set.Values(v.RelationID, v.Row)
+			if err != nil {
+				return fmt.Errorf("error parsing values: %s", err)
+			}
+
+			change := &types.Change{
+				Operation: "update",
+				Table:     changeSet.Name,
+				Fields:    make(map[string]interface{}, 0),
+				OldFields: make(map[string]interface{}, 0),
+			}
+			for name, value := range values {
+				change.Fields[name] = value.Get()
+			}
+
+			oldValues, err := set.Values(v.RelationID, v.OldRow)
+			if err == nil {
+				for name, value := range oldValues {
+					change.OldFields[name] = value
+				}
+			}
+
+			changeRecord.Changes = append(changeRecord.Changes, change)
+		case pgoutput.Delete:
+			changeSet, ok := set.Get(v.RelationID)
+			if !ok {
+				err := fmt.Errorf("relation not found for '%s'", changeSet.Name)
+				p.Log.Error(err)
+				return err
+			}
+			values, err := set.Values(v.RelationID, v.Row)
+			if err != nil {
+				values = make(map[string]pgtype.Value, 0)
+				p.Log.Debugf("Error parsing value: %s", err)
+			}
+
+			change := &types.Change{
+				Operation: "delete",
+				Table:     changeSet.Name,
+				Fields:    make(map[string]interface{}, 0),
+			}
+			for name, value := range values {
+				val := value.Get()
+
+				// Deletes will include the primary key with the rest of the fields being empty. Ignore them
+				if val == "" {
+					continue
+				}
+
+				change.Fields[name] = val
+			}
+			changeRecord.Changes = append(changeRecord.Changes, change)
+		}
+		return nil
+	}
+	return sub.Start(context.Background(), 0, handler)
+}
+
+// validateReadOptions ensures the correct CLI options are specified for the read action
+func validateReadOptions(opts *cli.Options) error {
+	return nil
```

Comment:
Do we not have to validate any inputs?

Reply:
Nope! they are already required at the CLI flag level. There is currently no optional/conditional inputs